ssl/ssl_sess.c

```diff
@@ -1251,4 +1268,168 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+        ERR_raise(ERR_LIB_SSL, SSL_R_BAD_CACHE_MODE);
+        return NULL;
+    }
+
+    if (!(mode & SSL_SESS_CACHE_NO_INTERNAL_LOOKUP)) {
+        /* Set up the search parameters */
+        memset(&data, 0, sizeof(data));
+        data.cache_id = s->cache_id;
+        data.cache_id_len = s->cache_id_len;
+        data.sid_ctx_length = s->sid_ctx_length;
+        if (s->sid_ctx_length > 0)
+            memcpy(data.sid_ctx, s->sid_ctx, s->sid_ctx_length);
+
+        /* search */
+        CRYPTO_THREAD_read_lock(s->session_ctx->lock);
+        ret = lh_SSL_SESSION_retrieve(s->session_ctx->sessions, &data);
+        if (ret != NULL)
+            SSL_SESSION_up_ref(ret);
+        CRYPTO_THREAD_unlock(s->session_ctx->lock);
+
+        if (ret == NULL) {
+            tsan_counter(&s->session_ctx->stats.sess_miss);
+        } else if (ret->timeout < (long)(time(NULL) - ret->time)) {
+            tsan_counter(&s->session_ctx->stats.sess_timeout);
+            SSL_CTX_remove_session(s->session_ctx, ret);
+            SSL_SESSION_free(ret);
+            ret = NULL;
+        } else {
+            tsan_counter(&s->session_ctx->stats.sess_hit);
+        }
+    }
+
+    if (ret == NULL && s->session_ctx->get_session_cb != NULL) {
+        int copy = 1;
+
+        ret = s->session_ctx->get_session_cb(s, s->cache_id,
+                                             s->cache_id_len,
+                                             &copy);
+
+        if (ret != NULL) {
+            tsan_counter(&s->session_ctx->stats.sess_hit);
+
+            /*
+             * Increment reference count now if the session callback asks us
+             * to do so (note that if the session structures returned by the
+             * callback are shared between threads, it must handle the
+             * reference count itself [i.e. copy == 0], or things won't be
+             * thread-safe).
+             */
+            if (copy)
+                SSL_SESSION_up_ref(ret);
+
+            /*
+             * Add the externally cached session to the internal cache as
+             * well if and only if we are supposed to.
+             */
+            if (!(mode & SSL_SESS_CACHE_NO_INTERNAL_STORE)) {
+                /*
+                 * Either return value of SSL_CTX_add_session should not
+                 * interrupt the session resumption process. The return
+                 * value is intentionally ignored.
+                 */
+                SSL_CTX_add_session(s->session_ctx, ret);
+            }
+        }
+    }
+
+    return ret;
+}
+
+int SSL_SESSION_set_timeout_update_cache(const SSL *s, long t)
+{
+    SSL_SESSION *ss, *ret;
+
+    ss = s->session;
+    if (ss == NULL)
+        return 0;
+
+    ss->timeout = t;
+
+    /* update cache in case it's a copy */
+    CRYPTO_THREAD_read_lock(s->session_ctx->lock);
+    ret = lh_SSL_SESSION_retrieve(s->session_ctx->sessions, ss);
+    if (ret != NULL && ret != ss)
+        ret->timeout = t;
```

Comment:
This looks problematic. The session object could be shared by multiple threads. Session really shout be immutable once created and put in the cache.

Reply:
`SSL_SESSION_set_time()` and `SSL_SESSION_set_timeout()` are used to modify the timeout of sessions in the cache.

Reply:
I agree with you that they should be immutable, but the timeouts can be modified.

Reply:
Access to the session timeout is serialized by the `SSL_CTX` lock, which is IMO compatible with the session being shared across threads.  The non-lifetime information remains immutable, so this is in a sense just adding some efficiency in terms of how to decide when an (immutable) session is no longer worth keeping around.